File: 2024/day_08/solution.py
from typing import List

def draw_race(indices: List[int], length: int) -> str:
  result = []

  for index, position in enumerate(indices):
    # Create the lane with '~' characters
    lane = ["~"] * length
    reindeer_position = position if position >= 0 else length + position

    # If the reindeer is not in the starting position, mark it with 'r'
    if position != 0:
      lane[reindeer_position] = "r"

    # Add the lane and the index of the reindeer to the result
    result.append(f"{' ' * (len(indices) - 1 - index)}{''.join(lane)} /{index + 1}")

  return "\n".join(result)

File: 2024/day_08/solution.js
function drawRace(indices, length) {
  return indices
    .map((position, index) => {
      const lane = Array(length).fill("~");
      const reindeerPosition = position >= 0 ? position : length + position;

      if (position !== 0) {
        lane[reindeerPosition] = "r";
      }

      return `${" ".repeat(indices.length - 1 - index)}${lane.join("")} /${
        index + 1
      }`;
    })
    .join("\n");
}

console.log(drawRace([0, 5, -3], 10));
console.log(drawRace([2, -1, 0, 5], 8));
console.log(drawRace([3, 7, -2], 12));
